### PR TITLE
Add (and use) durationFromSec and timeFromSec

### PR DESCRIPTION
**Pull request by dhood**
connects to ros2/navigation#5

I wanted to use tf2::timeFromSec and tf2::durationFromSec but they weren't implemented, so I added them in https://github.com/ros2/geometry2/commit/c7bac5b4d838e7de5900b3da200f256effbc615c (the code is copied [from this ros1 code](https://github.com/ros/roscpp_core/blob/057f7fa5c36df67757c6cb91c319faa67cefd6e6/rostime/include/ros/time.h#L153..L158))

In order to do implement those, (in particular [this line](https://github.com/ros2/geometry2/commit/c7bac5b4d838e7de5900b3da200f256effbc615c#diff-0ef30af0e2721678e4eb9da85f795979R70)) I made [this change to use `tf2::Duration` underneath `tf2::TimePoint`](https://github.com/ros2/geometry2/commit/958138e2fe11802c8771a580587a5b5c93adf3e4#diff-0ef30af0e2721678e4eb9da85f795979R43), so `double`s are used everywhere and not `long int`s from `std::chrono::nanoseconds` (thank you @Karsten1987 !!)

In doing this I noticed that with the exception of a few places (e.g. https://github.com/ros2/geometry2/commit/db4574a6564946a025630fc7cec49d2834cc689b#diff-473dd6a6d7f5086c80694ac302fd9f8eL137), we are accidentally initialising nanosecond durations instead of seconds. @clalancette idk if this is the underlying source of any of your TF issues.

There is still a question about why `double` is being used underneath `tf2::Duration` instead of the more standard `std::chrono::nanoseconds`. I think it was for 'nice' constructors e.g. `tf2::Duration(1.5)`, but that seems to have just led to mistakenly thinking it represented seconds. We could switch it or make a constructor that takes seconds.

### CODE REVIEW

**Comment by dhood** [comment]
note that for the executables to build properly this needs https://github.com/ros2/geometry2/pull/14

File: test_tf2/test/test_message_filter.cpp [hunk 1]
```diff
@@ -268,7 +268,7 @@ TEST(MessageFilter, outTheBackFailure)
 
   builtin_interfaces::msg::Time stamp(1);
   bc.setTransform(createTransform(Quaternion(0,0,0,1), Vector3(1,2,3), stamp, "frame1", "frame2"), "me");
-  bc.setTransform(createTransform(Quaternion(0,0,0,1), Vector3(1,2,3), stamp + tf2::Duration(10000), "frame1", "frame2"), "me");
+  bc.setTransform(createTransform(Quaternion(0,0,0,1), Vector3(1,2,3), stamp + tf2::durationFromSec(10000), "frame1", "frame2"), "me");
```
Comment (dhood) [inline test_tf2/test/test_message_filter.cpp:271]: if you're wondering if this was already ns, upstream it is also 10,000(s): https://github.com/ros/geometry2/blob/439e2350ca9b2f0581a3631f0929acc2021db73e/test_tf2/test/test_message_filter.cpp#L272

**Comment by dhood** [comment]
osx [Build Status](http://ci.ros2.org/job/ci_osx/2004/)
windows [Build Status](http://ci.ros2.org/job/ci_windows/2643/)
linux [Build Status](http://ci.ros2.org/job/ci_linux/2540/)
aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/139/)

**Comment by mikaelarguedas** [comment]
@tfoote to give feedback on this

**Comment by dhood** [comment]
I originally had it so that `timeFromSec` would give a time point offset from 'now'. I don't know what motivated that, but `ros::Time(t)` in ROS 1 [doesn't do that offset from 'now'](https://github.com/ros/roscpp_core/blob/057f7fa5c36df67757c6cb91c319faa67cefd6e6/rostime/include/ros/time.h#L153), so I have replicated that in https://github.com/ros2/geometry2/pull/16/commits/c622d9cc87e9c3d11b5c0e809dcbec634cd54745.

I went ahead and changed the storage of `tf2::TimePoint` to be integer and not double (https://github.com/ros2/geometry2/pull/16/commits/b4307d53b360611198386b7e1a48357e5fa2b94d). As @wjwwood pointed out, we shouldn't need double precision for measuring nanoseconds. I mentioned above this can lead to accidentally initialising Durations with ns when the user thinks the constructor takes seconds. Additionally, there are places where time comparisons will not work well with doubles e.g. [ this comparison of timestamps here](https://github.com/ros2/geometry2/blob/c5b4d4738ec1f824af364c9e55fb35239de3f1c6/tf2/src/cache.cpp#L112).

@clalancette could you please try this and let me know if you run into issues with this change? You may need to change some `tf2::Duration(t)` constructors to `tf2::durationFromSec(t)`.

**Comment by clalancette** [comment]
@dhood I'll give this a try today.

**Comment by dhood** [comment]
@clalancette no worries.

CI in the meantime: 
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/2555/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/153/)
* macOS  [Build Status](http://ci.ros2.org/job/ci_osx/2039/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/2664/)

File: tf2/include/tf2/time.h [hunk 2]
```diff
@@ -52,6 +53,23 @@ namespace tf2
     return std::chrono::system_clock::now();
   }
 
+  inline Duration durationFromSec(double t)
+  {
+    uint32_t sec, nsec;
+    sec = (uint32_t)floor(t);
+    nsec = (uint32_t)std::round((t-sec) * 1e9);
```
Comment (dhood) [inline tf2/include/tf2/time.h]: continued from https://github.com/ros2/geometry2/pull/20#discussion_r113841044 

@wjwwood wrote:
> Have you tried this function with `t_sec` < 0? I think you might want `std::trunc` rather than `std::floor`, and that makes me think that the nanoseconds logic might not be perfect either.
Maybe you could use this function's "possible implementation":
http://en.cppreference.com/w/cpp/chrono/duration/round
It is not available until C++17, but it's possible implementation might work in more cases than this.

I copied this code from ROS 1's implementation: https://github.com/ros/roscpp_core/blob/057f7fa5c36df67757c6cb91c319faa67cefd6e6/rostime/include/ros/time.h#L153..L158 but switched from using `boost::math::round` to `std::round`. the tests there only seem to test positive times  https://github.com/ros/roscpp_core/blob/057f7fa5c36df67757c6cb91c319faa67cefd6e6/rostime/test/time.cpp#L145 - is that something we should change?

I'll add similar tests here

Comment (wjwwood) [reply tf2/include/tf2/time.h]: I guess `round` actually makes sense here. It would still be nice to check the negative behavior in a test. My concern is that the result of `std::round((t-sec) * 1e9)` might be negative, which we are then just casting to an unsigned integer.

**Comment by dhood** [comment]
the switch to integer storage in https://github.com/ros2/geometry2/pull/16/commits/b4307d53b360611198386b7e1a48357e5fa2b94d broke the [durationToSec implementation](https://github.com/ros2/geometry2/blob/b4307d53b360611198386b7e1a48357e5fa2b94d/tf2/include/tf2/time.h#L74) (gets truncated to whole number). working on it.

**Comment by dhood** [comment]
I'm trying to update the `durationToSec` implementation to work with integer storage of `tf2::Duration`, modified in https://github.com/ros2/geometry2/commit/b4307d53b360611198386b7e1a48357e5fa2b94d

I'm stuck trying to get [this line](https://github.com/ros2/geometry2/pull/16/commits/16beb4e210b98b7df56f96c98f281a3a2e170f9b#diff-0ef30af0e2721678e4eb9da85f795979R79) to not turn an nsec integer value of, e.g. 639909744 into an nsec_double value of 0.63990974400000000255062104770332.

**Comment by mikaelarguedas** [comment]
Not sure there is a good way to work around rounding errors (at least in C). I guess the easiest would be to make sure that the "diff" is lower that your "precision" aka 1e-9. Also I think there are few systems that can be timestamping precisely at that scale

**Comment by dhood** [comment]
Here's where I'm at with the updated implementation of `durationToSec`.

Given a chrono duration, from its `count` in ns, the seconds and nanoseconds parts can be calculated individually quite accurately as doubles. It is [combining them with addition](https://github.com/ros2/geometry2/pull/16/commits/252615a45050e75078ed92b5d6b2ea657b833e88#diff-0ef30af0e2721678e4eb9da85f795979R83) that introduces the floating point error. Example:
```
3: Count (ns): 1494614995898323272
3: sec_double: 1494614995.0000000000000000
3: nsec_double: 0.8983232720000001
3: Returning double of: 1494614995.8983232975006104
3: Difference to expected (ns): 26
```

This seems like reasonable behaviour, people shouldn't expect the double representation of a timepoint to be exact. However, a tolerance of 1e-9 is too strict: the value can be off by up to ~150ns.

it doesn't seem to work properly for negatives so I'm working on that now..

**Comment by dhood** [comment]
Back in review.

I have added tests for the new functions, which now include testing negative durations: https://github.com/ros2/geometry2/blob/978b8083bf2e64ed3b96288af344fc929445d338/tf2/test/simple_tf2_core.cpp#L159..L212 (using pre-defined values [here](https://github.com/ros2/geometry2/blob/978b8083bf2e64ed3b96288af344fc929445d338/tf2/test/simple_tf2_core.cpp#L184) because I understood from [this comment in the ros1 code](https://github.com/ros/roscpp_core/blob/057f7fa5c36df67757c6cb91c319faa67cefd6e6/rostime/test/time.cpp#L41) that random values aren't ideal).

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/2673/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/223/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/2129/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/2792/) (unrelated flaky test)

**Comment by dhood** [comment]
@ros2/team this is in review (wasn't showing on waffle board)

To check that the new tests aren't flaky, these jobs have test args of `--only tf2 --retest-until-fail 20`:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/2691/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/235/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/2150/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/2815/)

**Comment by dhood** [comment]
still waiting for review

corresponding update to cartographer in https://github.com/ros2/cartographer_ros/pull/3
anything else that breaks from the change from double to integer storage was probably intending to use seconds not nanoseconds (e.g. [all of these updated tests](https://github.com/ros2/geometry2/pull/16/commits/0090e41ac8c031a687a11c975b4635a2c8f0ab46)) and so wouldn't be working as expected anyway.

File: tf2/include/tf2/time.h [hunk 3]
```diff
@@ -52,8 +54,32 @@ namespace tf2
     return std::chrono::system_clock::now();
   }
 
-  inline double durationToSec(const tf2::Duration & input){
-    return (double)std::chrono::duration_cast<std::chrono::seconds>(input).count();
+  inline Duration durationFromSec(double t_sec)
+  {
+    int32_t sec, nsec;
+    sec = (int32_t)floor(t_sec);
+    nsec = (int32_t)std::nearbyint((t_sec-sec) * 1e9);
```
Comment (wjwwood) [inline tf2/include/tf2/time.h]: Please use C++ style casts.

https://stackoverflow.com/a/1609185/671658

File: tf2/include/tf2/time.h [hunk 4]
```diff
@@ -52,8 +54,32 @@ namespace tf2
...
+  inline TimePoint timeFromSec(double t_sec)
+  {
+    return tf2::TimePoint(durationFromSec(t_sec));
+  }
+
+  inline double durationToSec(const tf2::Duration& input){
+    int64_t count = input.count();
+    int32_t sec, nsec;
+    nsec = (int32_t)(count % 1000000000l);
+    sec = (int32_t)((count - nsec) / 1000000000l);
```
Comment (wjwwood) [inline tf2/include/tf2/time.h]: Please use C++ style casts.

**Review by wjwwood — CHANGES_REQUESTED** [review]
lgtm, except for casting style being used in two places (that I saw)

File: tf2/include/tf2/time.h (hunk 3 above)
Comment (wjwwood) [reply tf2/include/tf2/time.h]: Btw, this isn't in our style guide, but it is in the Google Style guide which ours is based on: https://google.github.io/styleguide/cppguide.html#Casting

When writing our style guide I tried to avoid copying the entire Google Style guide, so I don't know if this can be made more visible or not.

I also do not think cpplint.py checks this (it may not be possible to do so easily).

Comment (dhood) [reply tf2/include/tf2/time.h]: ok, will do! (was copying from ROS 1 code not ROS 2 code)

Comment (mikaelarguedas) [reply tf2/include/tf2/time.h]: > I also do not think cpplint.py checks this (it may not be possible to do so easily).

IIRC cpplint does check this, but I don't think we run linter on this package

**Comment by dhood** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/2712/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/254/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/2172/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/2842/)

File: tf2/include/tf2/time.h [hunk 5]
```diff
@@ -52,8 +54,32 @@ namespace tf2
     return std::chrono::system_clock::now();
   }
 
-  inline double durationToSec(const tf2::Duration & input){
-    return (double)std::chrono::duration_cast<std::chrono::seconds>(input).count();
+  inline Duration durationFromSec(double t_sec)
+  {
+    int32_t sec, nsec;
+    sec = static_cast<int32_t>(floor(t_sec));
+    nsec = static_cast<int32_t>(std::nearbyint((t_sec-sec) * 1e9));
```
Comment (clalancette) [inline tf2/include/tf2/time.h]: Do we need to explicitly set the rounding mode for std::nearbyint?

Comment (clalancette) [reply tf2/include/tf2/time.h]: Also, nit: spaces around the minus sign.

**Comment by dhood** [comment]
new CI after review
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/2716/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/259/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/2181/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/2848/)

**Review by wjwwood — APPROVED** [review]
